### PR TITLE
Testdox improvements

### LINKED ISSUE #3567: Unbuffered TestDox printer option
_As a_ developer of slower integration tests
_I want to_ see prettified tests results as soon as tests are completed
_so that_ my workflow is both agile and æsthetic

## Background ##
The [buffering testdox printer](#3417) and its friend [the spinner](#3513) are [breaking some workflows](https://github.com/sebastianbergmann/phpunit/issues/3560#issuecomment-473871720) that people used to good effect.

The example from #3560 is using Testdox to get readable results while running. The actual test order is not that important, set by configuration or implied by test collection structure. My implementation of the buffering printer breaks this specific use case.

## Solution ##
Implement explicit "do not buffer result ouput" option, along the lines of #3513:
- [x] extend buffering mechanism to explicitly support always-flush
- [x] add CLI flag
- [x] add end-to-end tests voor `--enable-buffer` and `--disable-buffer` flag
- [x] add XML-configuration option
- [x] add unit test coverage for unbuffered Testdox

### LINKED ISSUE #3900: Reduce the number of lines in Testdox output when using a data provider
**Current situation/problem:**

We use dataproviders with a big number of data sets in our unit tests. For each data set PHPUnit prints a line in the testdox output. In our case this is overwhelming and makes the testdox output very hard to read and therefore useless for the product owner.

It is the very high number of output lines (not the wording) that make the Testdox output unclear. At the moment it may not be usable for product owners if you use dataproviders with a big number of data sets.

**Feature request:**

Would be great, if there was a possibility to let PHPUnit print only one line per test (regardless of the number of data sets the dataprovider returns) in the testdox output.

Possible solutions:

- a global configuration setting that effects the testdox output of all tests
- a annotation that only effects the testdox output of a specific test

**Example**

This is a snippet of our current Testdox output:

```
 ...
 ✔ Returns the gross amount to be used for the account for all appropriate accounting records with data set 805
 ✔ Returns the gross amount to be used for the account for all appropriate accounting records with data set 806
 ✔ Returns the gross amount to be used for the account for all appropriate accounting records with data set 807
 ✔ Returns the gross amount to be used for the account for all appropriate accounting records with data set 808
 ✔ Returns the gross amount to be used for the account for all appropriate accounting records with data set 809
 ✔ Returns the gross amount to be used for the account for all appropriate accounting records with data set 810
 ✔ Returns the gross amount to be used for the account for all appropriate accounting records with data set 811
 ✔ Returns the gross amount to be used for the account for all appropriate accounting records with data set 812
 ✔ Returns the gross amount to be used for the account for all appropriate accounting records with data set 813
 ✔ Returns the gross amount to be used for the account for all appropriate accounting records with data set 814
 ✔ Returns the gross amount to be used for the account for all appropriate accounting records with data set 815
 ✔ Returns the net amount to be used for the contra account for taxable aktivkonto an aufwandkonto accounting records
 ✔ Returns the net amount to be used for the contra account for taxable aktivkonto an ertragkonto accounting records
 ✔ Returns the net amount to be used for the contra account for taxable kundenkonto an aktivkonto accounting records
 ✔ Returns the net amount to be used for the contra account for taxable kundenkonto an ertragkonto accounting records
 ✔ Returns the net amount to be used for the contra account for taxable lieferantenkonto an aufwandkonto accounting records
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records with data set 0
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records with data set 1
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records with data set 2
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records with data set 3
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records with data set 4
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records with data set 5
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records with data set 6
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records with data set 7
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records with data set 8
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records with data set 9
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records with data set 10
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records with data set 11
 ...
```

This could be a more condensed output e.g. if no data set would let a test fail:

```
...
 ✔ Returns the gross amount to be used for the account for all appropriate accounting records
 ✔ Returns the net amount to be used for the contra account for taxable aktivkonto an aufwandkonto accounting records
 ✔ Returns the net amount to be used for the contra account for taxable aktivkonto an ertragkonto accounting records
 ✔ Returns the net amount to be used for the contra account for taxable kundenkonto an aktivkonto accounting records
 ✔ Returns the net amount to be used for the contra account for taxable kundenkonto an ertragkonto accounting records
 ✔ Returns the net amount to be used for the contra account for taxable lieferantenkonto an aufwandkonto accounting records
 ✔ Returns the gross amount to be used for the contra account for all appropriate accounting records
 ...
```

### PR DESCRIPTION
**Pull request by epdenouden**
- [x] implements #3567 ability to toggle buffering
- [ ] implements #3900 reduce data provider noise
- [x] improve unit test coverage in general
  - [x] Testdox buffering mechanism as seperate from colorization
  - [x] colorized Testdox features like the spinner
  - [x] more coverage Configuration value object
- [x] bunch of small fixes (typos, code style)

### CODE REVIEW

File: src/Util/TestDox/TestDoxPrinter.php [hunk 1]
```diff
@@ -96,6 +128,11 @@ public function setOriginalExecutionOrder(array $order): void
         $this->enableOutputBuffer     = !empty($order);
     }
 
+    public function setEnableOutputBuffer(bool $enabled = true): void
```
Comment (localheinz) [inline src/Util/TestDox/TestDoxPrinter.php:131]: Since the default is `false`, perhaps we only need a method `enableOutputBuffer()`, and maybe `disableOutputBuffer()`, both without parameters?

File: src/Util/TestDox/TestDoxPrinter.php [hunk 2]
```diff
@@ -183,6 +220,25 @@ public function flush(): void
         $this->flushOutputBuffer(true);
     }
 
+    public function startTestSuite(TestSuite $suite): void
+    {
+        parent::startTestSuite($suite); // TODO: Change the autogenerated stub
```
Comment (localheinz) [inline src/Util/TestDox/TestDoxPrinter.php:225]:
```suggestion
        parent::startTestSuite($suite);
```

File: src/Util/TestDox/TestDoxPrinter.php [hunk 3]
```diff
@@ -183,6 +220,25 @@ public function flush(): void
...
+
+        $this->currentSuiteName           = $suite->sortId();
+        $this->currentSuiteIsDataProvider = $suite instanceof DataProviderTestSuite;
+        $this->testSuiteStack[]           = $this->currentSuiteName;
+    }
+
+    public function endTestSuite(TestSuite $suite): void
+    {
+        parent::endTestSuite($suite); // TODO: Change the autogenerated stub
```
Comment (localheinz) [inline src/Util/TestDox/TestDoxPrinter.php:234]:
```suggestion
        parent::endTestSuite($suite);
```

File: src/Util/TestDox/TestDoxPrinter.php [hunk 4]
```diff
@@ -191,21 +247,23 @@ protected function registerTestResult(Test $test, ?Throwable $t, int $status, fl
...
-            'verbose'    => $verbose,
+            'className'    => $this->formatClassName($test),
+            'testName'     => $testName,
+            'testMethod'   => $this->formatTestName($test),
+            'message'      => '',
+            'status'       => $status,
+            'time'         => $time,
+            'verbose'      => $verbose,
+            'suite'        => $this->currentSuiteName,
+            'dataprovider' => $this->currentSuiteIsDataProvider,
```
Comment (localheinz) [inline src/Util/TestDox/TestDoxPrinter.php:258]: What do you think about using consistent casing?

```suggestion
            'dataProvider' => $this->currentSuiteIsDataProvider,
```

File: src/Util/TestDox/TestDoxPrinter.php [hunk 5]
```diff
@@ -233,45 +291,55 @@ protected function testHasPassed(): bool
...
-                    $this->testFlushIndex++;
-                    $prevResult = $result;
-                    $flushed    = true;
-                } else {
+        // Buffered output: match original test load order as close as possible
+        do {
+            if (isset($this->originalExecutionOrder[$this->resultFlushCount])) {
+                $result = $this->getTestResultByName($this->originalExecutionOrder[$this->resultFlushCount]);
+
+                if (empty($result)) {
```
Comment (localheinz) [inline src/Util/TestDox/TestDoxPrinter.php:327]:
```suggestion
                if ([] === $result) {
```

File: src/Util/TestDox/TestDoxPrinter.php [hunk 6]
```diff
@@ -233,45 +291,55 @@ protected function testHasPassed(): bool
...
                     $this->showSpinner();
+
+                    return;
                 }
-            } while ($flushed && $this->testFlushIndex < $this->testIndex);
-        }
+            } else {
+                // Edge case for exceptions in tearDownAfterClass
+                // This test(name) cannot found in original execution order,
```
Comment (localheinz) [inline src/Util/TestDox/TestDoxPrinter.php:334]:
```suggestion
                // This test(name) cannot be found in original execution order,
```

File: tests/unit/Util/TestDox/TestDoxPrinterTest.php [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+     *
+     * @return array<string>
+     */
+    private function getOutputLinesForTestOrder(?array $testNames = null): array
+    {
+        if ($testNames === null) {
+            $testNames = $this->originalOrder;
+        }
+
+        return array_map(static function (string $t) {
```
Comment (localheinz) [inline tests/unit/Util/TestDox/TestDoxPrinterTest.php:289]:
```suggestion
        return array_map(static function (string $t): string {
```

File: src/TextUI/CliArguments/Configuration.php [hunk 8]
```diff
@@ -437,10 +437,15 @@ final class Configuration
...
...rray $extensions, ?array $unavailableExtensions, ?bool $failOnEmptyTestSuite, ?bool $failOnIncomplete, ?bool $failOnRisky, ?bool $failOnSkipped, ?bool $failOnWarning, ?string $filter, ?bool $generateConfiguration, ?array $groups, ?bool $help, ?string $includePath, ?array $iniSettings, ?string $junitLogfile, ?bool $listGroups, ?bool $listSuites, ?bool $listTests, ?string $listTestsXml, ?string $loader, ?bool $noCoverage, ?bool $noExtensions, ?bool $noInteraction, ?bool $noLogging, ?string $printer, ?bool $processIsolation, ?int $randomOrderSeed, ?int $repeat, ?bool $reportUselessTests, ?bool $resolveDependencies, ?bool $reverseList, ?bool $stderr, ?bool $strictCoverage, ?bool $stopOnDefect, ?bool $stopOnError, ?bool $stopOnFailure, ?bool $stopOnIncomplete, ?bool $stopOnRisky, ?bool $stopOnSkipped, ?bool $stopOnWarning, ?string $teamcityLogfile, ?array $testdoxExcludeGroups, ?array $testdoxGroups, ?string $testdoxHtmlFile, ?string $testdoxTextFile, ?string $testdoxXmlFile, ?array $testSuffixes, ?string $testSuite, array $unrecognizedOptions, ?string $unrecognizedOrderBy, ?bool $useDefaultConfiguration, ?bool $verbose, ?bool $version, ?array $coverageFilter, ?string $xdebugFilterFile)
```
[the remark this answers is not in the document]
Comment (epdenouden) [reply src/TextUI/CliArguments/Configuration.php:448]: I think we'd better just buy some extra disinfectant ;)

File: src/TextUI/TestRunner.php [hunk 9]
```diff
@@ -1103,6 +1105,7 @@ private function handleConfiguration(array &$arguments): void
         $arguments['disallowTestOutput']                              = $arguments['disallowTestOutput'] ?? false;
         $arguments['disallowTodoAnnotatedTests']                      = $arguments['disallowTodoAnnotatedTests'] ?? false;
         $arguments['defaultTimeLimit']                                = $arguments['defaultTimeLimit'] ?? 0;
+        $arguments['enableOutputBuffer']                              = $arguments['enableOutputBuffer'] ?? true;
```
[the remark this answers is not in the document]
Comment (epdenouden) [reply src/TextUI/TestRunner.php:1108]: The whole `$arguments` handling could use some care and attention, no argument there.

File: src/Util/Printer.php [hunk 10]
```diff
@@ -92,6 +92,10 @@ public function __construct($out = null)
 
     public function write(string $buffer): void
     {
+        if ($buffer === '') {
```
[the remark this answers is not in the document]
Comment (epdenouden) [reply src/Util/Printer.php:95]: This is just a little change to make testing easier. Without there would be a bunch of unrelated and useless empty write events in the buffer. 

Also calling the writer with an empty piece of paper is some next-level _Bonpflicht_ ;-)

File: src/Util/TestDox/CliTestDoxPrinter.php [hunk 11]
```diff
@@ -181,22 +181,22 @@ protected function formatTestName(Test $test): string
         return parent::formatTestName($test);
     }
 
-    protected function writeTestResult(array $prevResult, array $result): void
+    protected function writeSingleTestResult(array $result): void
```
[the remark this answers is not in the document]
Comment (epdenouden) [reply src/Util/TestDox/CliTestDoxPrinter.php:184]: I will double-check and restore the older methodname as a proxy for the newer one. There will be new ones for the folding of passing data providers and test suites, like e.g. `writeCompletedDataProviderSuite`

[the remark this answers is not in the document]
Comment (epdenouden) [reply src/Util/TestDox/CliTestDoxPrinter.php:184]: To clarify: the `writeSingleTestResult` is meant for internal TestDox use only. Until we switch to the event system I am not going to be overly worried about BC-breaks for monkey-patched extensions.

File: src/Util/TestDox/TestDoxPrinter.php [hunk 12]
```diff
@@ -40,12 +46,12 @@ class TestDoxPrinter extends DefaultResultPrinter
     /**
      * @var int The number of test results received from the TestRunner
      */
-    protected $testIndex = 0;
+    protected $resultCount = 0;
```
[the remark this answers is not in the document]
Comment (epdenouden) [reply src/Util/TestDox/TestDoxPrinter.php:49]: It only 'breaks' between our own TestDox printer classes, for which it was a simple name change. I seriously hope nobody out there is patching my colorful monstrosity

File: src/Util/TestDox/TestDoxPrinter.php [hunk 13]
```diff
@@ -40,12 +46,12 @@ class TestDoxPrinter extends DefaultResultPrinter
...
 
     /**
      * @var int The number of test results already sent to the output
      */
-    protected $testFlushIndex = 0;
+    protected $resultFlushCount = 0;
```
[the remark this answers is not in the document]
Comment (epdenouden) [reply src/Util/TestDox/TestDoxPrinter.php:54]: yes, same as the other naming changes in the Testdox classes

File: src/Util/TestDox/TestDoxPrinter.php (hunk 1 above)
Comment (epdenouden) [reply src/Util/TestDox/TestDoxPrinter.php:131]: Good point. There was some obscure reason I hadn't done that yet. Will check my notes.

**Comment by Jellyfrog** [comment]
Maybe this could be splitted into multiple PRs to speed up time to market?

**Comment by nreynis** [comment]
> Maybe this could be splitted into multiple PRs to speed up time to market?

Yes please, the disable-buffer is a crucial feature if you are doing continuous fullstack testing.  
Without it testdox is reallly crippled : any test that goes wrong or make a job timeout and you went from having usable log to no log at all when you needed them the most.

**Comment by sebastianbergmann** [comment]
Thank you for your contribution. I appreciate the time you invested in preparing this pull request. However, I have decided not to merge it.
